### PR TITLE
Cdp extend

### PR DESCRIPTION
**Pull request by poliwop**
Added money market to model liquidations

### CODE REVIEW

**Review by jepidoptera — CHANGES_REQUESTED** [review]
One general issue I would raise is that using "raise" without indicating a specific exception really isn't best practice, especially when considering future users who may be less familiar with the codebase. I don't think that's a blocking issue but I don't think it's worth considering. Mostly this looks good and pretty thorough, but see the comments.

File: hydradx/model/amm/liquidations.py [hunk 1]
```diff
@@ -1,34 +1,181 @@
...
+            self.liquidation_penalty = {asset: liquidation_penalty for asset in liquidity}
+
+        if full_liquidation_threshold is None:
+            self.full_liquidation_threshold = {key: value for key, value in self.liquidation_threshold.items()}
+        elif isinstance(full_liquidation_threshold, dict):
+            self.full_liquidation_threshold = full_liquidation_threshold
+        else:
+            self.full_liquidation_threshold = {asset: full_liquidation_threshold for asset in liquidity}
+
+        self.oracles = oracles  # MoneyMarket should never mutate the oracles
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/liquidations.py:63]: That's a good point, I would like to make this change in a future update (I am already making some extensions in a new branch so I will make the change in that branch).

File: hydradx/model/amm/global_state.py [hunk 2]
```diff
@@ -426,32 +435,56 @@ def find_partial_liquidation_amount(omnipool: OmnipoolState, cdp: CDP, penalty:
     return delta_debt_down
 
 
-def omnipool_liquidate_cdp(omnipool: OmnipoolState, cdp: CDP, treasury_agent: Agent, delta_debt: float, penalty: float) -> None:
+def omnipool_liquidate_cdp(omnipool: OmnipoolState, mm: money_market, cdp_i: int, treasury_agent: Agent,
+                           delta_debt: float) -> None:
+    cdp = mm.cdps[cdp_i][1]
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/global_state.py:440]: I am not sure about this, but I agree it's confusing that some use cdp: CDP and some use cdp_i: int.

Some functions make sense outside the context of a money market with an arbitrary CDP, or even applying some parameters of a money market to a CDP that isn't in that money market. For others, you want to enforce that you are the function only applies to CDPs in the money market.

I think however the main use case for having cdp: CDP is to use functions on "fictional" versions of CDPs in the money market, essentially to do "what-if" calculations.

I'll think about if I can do cdp: CDP with everything. Maybe I'll only retain cdp_i: int if it's absolutely necessary that the cdp in question be in the current money market.

File: hydradx/model/amm/liquidations.py [hunk 3]
```diff
@@ -1,34 +1,181 @@
...
+        conversion = (1 + self.liquidation_penalty[cdp.collateral_asset]) / self.get_oracle_price(cdp.collateral_asset,
+                                                                                                  cdp.debt_asset)
+        actual_debt_liq = self._calculate_debt(cdp, delta_debt)
+        return actual_debt_liq * conversion
+
+    def borrow(self, agent: Agent, borrow_asset: str, collateral_asset: str, borrow_amt: float,
+               collateral_amt: float) -> None:
+        assert borrow_asset != collateral_asset
+        assert borrow_asset in self.liquidity
+        assert borrow_amt <= self.liquidity[borrow_asset]
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/liquidations.py:116]: I think you meant it should say
"assert borrow_amt <= self.liquidity[borrow_asset] **-** self.borrowed[borrow_asset]"
right?

File: hydradx/model/amm/liquidations.py [hunk 4]
```diff
@@ -1,34 +1,181 @@
...
+        assert agent.is_holding(collateral_asset, collateral_amt)
+        if (borrow_asset, collateral_asset) in self.oracles:
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/liquidations.py:118]: Agreed

File: hydradx/tests/test_liquidations.py [hunk 5]
```diff
@@ -32,69 +35,363 @@ def test_cdp_validate():
...
+    with pytest.raises(Exception):  # should fail because LTV is too low
+        mm.borrow(agent, borrow_asset, collateral_asset, borrow_amt, collat_amt)
+    borrow_amt = collat_amt * 10 * 0.50
+    with pytest.raises(Exception):  # should fail because debt asset == collateral asset
+        mm.borrow(agent, collateral_asset, collateral_asset, borrow_amt, collat_amt)
+    with pytest.raises(Exception):  # should fail because collateral asset is not in agent holdings
+        mm.borrow(agent, borrow_asset, "ETH", borrow_amt, collat_amt)
+    with pytest.raises(Exception):  # should fail because borrow asset is not in liquidity
+        mm.borrow(agent, "ETH", collateral_asset, borrow_amt, collat_amt)
+    with pytest.raises(Exception):  # should fail because of missing oracle
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/tests/test_liquidations.py:337]: Agreed

File: hydradx/model/amm/liquidations.py [hunk 6]
```diff
@@ -1,34 +1,181 @@
...
 
     def validate(self) -> bool:
         if self.debt_amt < 0 or self.collateral_amt < 0:
             return False
         if self.debt_asset == self.collateral_asset:
             return False
         return True
+
+
+class money_market:
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/liquidations.py:33]: Agreed, I'd like to fix this is the next PR

File: hydradx/model/amm/global_state.py [hunk 7]
```diff
@@ -372,52 +369,64 @@ def transform(state: GlobalState) -> GlobalState:
...
+    """Find largest amount of debt from a CDP that can be liquidated profitably against Omnipool.
+
+    Args:
+        omnipool: Omnipool state
+        mm: money market
+        cdp_i: int index of CDP in mm.cdps list
+        iters: max iterations for the binomial search (default 20)
+        min_amt: minimum debt amount to liquidate (default 0)
+    """
+    cdp = mm.cdps[cdp_i][1]
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/global_state.py:398]: Agreed

**Review by jepidoptera — APPROVED** [review]

**Pull request merged**